### PR TITLE
Merkle db iterator

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
Add a proper iterator to the trieview.  Improves performance of proof generation by reducing the number of keys loaded in the underlying trie by being more aware of the new keys in the view.  Followup will be to implement the rest of the DB interface in trieview.

### CODE REVIEW

File: x/merkledb/view_iterator.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	changes := make([]KeyChange, 0, len(t.changes.values))
+	for path, change := range t.changes.values {
+		key := path.Serialize().Value
+		if (len(start) > 0 && bytes.Compare(start, key) > 0) || !bytes.HasPrefix(key, prefix) {
+			continue
+		}
+		changes = append(changes, KeyChange{Key: key, Value: change.after})
+	}
+	// sort [changes] so they can be merged with the parent trie's state
+	slices.SortFunc(changes, func(a, b KeyChange) bool {
```
Comment (dboehm-avalabs) [inline x/merkledb/view_iterator.go:35]: We could theoretically store these sorted all the time, but I'm not sure if it is worth it.

File: x/merkledb/view_iterator_test.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+	require.False(iterator.Next())
+	require.Nil(iterator.Key())
+	require.Nil(iterator.Value())
+	require.NoError(iterator.Error())
+}
+
+// Test_TrieView_IteratorMemorySafety tests to make sure that keys and values can
+// be modified from the returned iterator.
+func Test_TrieView_IteratorMemorySafety(t *testing.T) {
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:153]: doesn't seem this test modifies the returned key/values?

File: x/merkledb/trieview.go [hunk 3]
```diff
@@ -418,26 +420,18 @@ func (t *trieView) GetRangeProof(
...
+	result.KeyValues = make([]KeyValue, 0, initKeyValuesSize)
+	it := t.NewIteratorWithStart(start)
+	for it.Next() && len(result.KeyValues) < maxLength && (len(end) == 0 || bytes.Compare(it.Key(), end) <= 0) {
+		// clone the value to prevent editing of the values stored within the trie
+		result.KeyValues = append(result.KeyValues, KeyValue{
+			Key:   it.Key(),
+			Value: slices.Clone(it.Value()),
+		})
 	}
+	it.Release()
```
Comment (darioush) [inline x/merkledb/trieview.go:434]: should we check `it.Error()`?

File: x/merkledb/view_iterator.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		case len(it.sortedChanges) == 0:
+			it.key = it.parentIter.Key()
+			it.value = it.parentIter.Value()
+			it.parentIterExhausted = !it.parentIter.Next()
+			return true
+		default:
+			memKey := it.sortedChanges[0].Key
+			memValue := it.sortedChanges[0].Value
+
+			dbKey := it.parentIter.Key()
```
Comment (darioush) [inline x/merkledb/view_iterator.go:116]: nit: should we rename this to `parentKey` so it matches the comments?

File: x/merkledb/view_iterator_test.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		if len(expectedValue) == 0 {
+			// Don't differentiate between nil and []byte{}
+			require.Empty(iter.Value(), 0)
+		} else {
+			require.Equal(expectedValue, iter.Value())
+		}
+		i++
+	}
+	require.Equal(len(uniqueKeys), i)
+	iter.Release()
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:232]: should we check `it.Error()`?

File: x/merkledb/view_iterator_test.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	uniqueKeys := maps.Keys(uniqueKeyChanges)
+	sort.Strings(uniqueKeys)
+	i := 0
+	for iter.Next() {
+		expectedKey := uniqueKeys[i]
+		expectedValue := uniqueKeyChanges[expectedKey]
+		require.Equal([]byte(expectedKey), iter.Key())
+		if len(expectedValue) == 0 {
+			// Don't differentiate between nil and []byte{}
+			require.Empty(iter.Value(), 0)
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:225]:
```suggestion
			require.Empty(iter.Value())
```

File: x/merkledb/view_iterator_test.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+		} else {
+			require.Equal(expectedValue, iter.Value())
+		}
+		i++
+	}
+	require.Equal(len(uniqueKeys), i)
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:231]:
```suggestion
	require.Len(uniqueKeys, i)
```

File: x/merkledb/view_iterator_test.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+		require.Equal([]byte(expectedKey), iter.Key())
+		if len(expectedValue) == 0 {
+			// Don't differentiate between nil and []byte{}
+			require.Empty(iter.Value())
+		} else {
+			require.Equal(expectedValue, iter.Value())
+		}
+		i++
+	}
+	require.Equal(len(startPrefixUniqueKeys), i)
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:259]:
```suggestion
	require.Len(startPrefixUniqueKeys, i)
```

File: x/merkledb/view_iterator_test.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	iter.Release()
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:260]: nit: should we check `it.Error()`

File: x/merkledb/view_iterator_test.go [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+	start := []byte{128, 5}
+	iter = view3.NewIteratorWithStartAndPrefix(start, prefix)
+	startPrefixUniqueKeys := []string{}
+	// Remove keys that don't have the prefix/are before the start.
+	for i := 0; i < len(uniqueKeys); i++ {
+		if bytes.HasPrefix([]byte(uniqueKeys[i]), prefix) && bytes.Compare([]byte(uniqueKeys[i]), start) >= 0 {
+			startPrefixUniqueKeys = append(startPrefixUniqueKeys, uniqueKeys[i])
+		}
+	}
+	require.Greater(len(startPrefixUniqueKeys), 0) // Sanity check to make sure we have some keys to test.
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:245]:
```suggestion
	require.NotEmpty(startPrefixUniqueKeys) // Sanity check to make sure we have some keys to test.
```

File: x/merkledb/view_iterator_test.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+	i = 0
+	for iter.Next() {
```
Comment (darioush) [inline x/merkledb/view_iterator_test.go:247]: should we make this a fn that we can call twice?

File: x/merkledb/view_iterator.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+	err        error
+
+	sortedChanges []KeyChange
+
+	initialized, parentIterExhausted bool
+}
+
+// Next moves the iterator to the next key/value pair. It returns whether the
+// iterator is exhausted. We must pay careful attention to set the proper values
+// based on if the in memory db or the underlying db should be read next
```
Comment (darioush) [inline x/merkledb/view_iterator.go:67]:
```suggestion
// based on if the in memory changes or the underlying db should be read next
```

**Review by darioush — APPROVED** [review]

**Review by danlaine — APPROVED** [review]

**Pull request merged**